File: hf_dataset_download.py
#!/usr/bin/env python3
"""交互式下载 Hugging Face Dataset Viewer 中的数据为 JSONL。"""

import argparse
import json
import os
import re
import sys
from pathlib import Path
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen


API_BASE = "https://datasets-server.huggingface.co"
PAGE_SIZE = 100  # Dataset Viewer API 单次最多返回 100 条。


def choose(prompt: str, options: list[str], default: int = 0) -> int:
    """显示选项，返回用户选择的下标。"""
    print(f"\n{prompt}")
    for index, option in enumerate(options, start=1):
        suffix = "（默认）" if index == default + 1 else ""
        print(f"  {index}. {option}{suffix}")

    while True:
        answer = input(f"请输入编号 [默认 {default + 1}]: ").strip()
        if not answer:
            return default
        try:
            index = int(answer) - 1
            if 0 <= index < len(options):
                return index
        except ValueError:
            pass
        print("输入无效，请输入列表中的编号。")


def get_token() -> str | None:
    """优先使用环境变量，其次读取 `hf auth login` 保存的 Token。"""
    if token := os.environ.get("HF_TOKEN"):
        return token

    token_path = os.environ.get("HF_TOKEN_PATH")
    if token_path:
        paths = [Path(token_path)]
    else:
        cache_root = os.environ.get("XDG_CACHE_HOME", str(Path.home() / ".cache"))
        hf_home = Path(os.environ.get("HF_HOME", Path(cache_root) / "huggingface"))
        paths = [hf_home / "token"]

    for path in paths:
        try:
            if token := path.read_text(encoding="utf-8").strip():
                return token
        except OSError:
            continue
    return None


def request_json(path: str, params: dict[str, Any], timeout: int) -> dict[str, Any]:
    url = f"{API_BASE}{path}?{urlencode(params)}"
    headers = {"User-Agent": "hf-dataset-sample-downloader/1.0"}
    if token := get_token():
        headers["Authorization"] = f"Bearer {token}"

    try:
        with urlopen(Request(url, headers=headers), timeout=timeout) as response:
            return json.load(response)
    except HTTPError as error:
        detail = error.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"HTTP {error.code}: {detail}") from error
    except URLError as error:
        raise RuntimeError(f"网络请求失败：{error.reason}") from error


def safe_name(text: str) -> str:
    return re.sub(r"[^A-Za-z0-9_.-]+", "_", text)


def main() -> int:
    parser = argparse.ArgumentParser(
        description="选择 Hugging Face 数据集的 split，并下载 100、1k 或全部记录。"
    )
    parser.add_argument("dataset", help="数据集 ID，例如 HuggingFaceH4/ultrachat_200k")
    parser.add_argument("--config", help="只显示指定配置（subset）下的 split")
    parser.add_argument("-o", "--output", help="输出 JSONL 文件路径")
    parser.add_argument("--timeout", type=int, default=60, help="单次网络请求超时秒数，默认 60")
    args = parser.parse_args()

    try:
        split_response = request_json("/splits", {"dataset": args.dataset}, args.timeout)
        splits = split_response.get("splits", [])
        if args.config:
            splits = [item for item in splits if item.get("config") == args.config]
        if not splits:
            config_tip = f"（config: {args.config}）" if args.config else ""
            raise RuntimeError(f"没有找到可用 split {config_tip}")

        labels = []
        for item in splits:
            total = item.get("num_rows")
            total_text = f"，{total:,} 条" if isinstance(total, int) else ""
            labels.append(f"{item['config']} / {item['split']}{total_text}")
        selected = splits[choose("可用 config / split：", labels)]

        amount_labels = ["100 条", "1k 条", "全部"]
        amount = amount_labels[choose("下载数量：", amount_labels)]
        requested = {"100 条": 100, "1k 条": 1000, "全部": None}[amount]
        available = selected.get("num_rows")
        if requested is None:
            target = available if isinstance(available, int) else None
        elif isinstance(available, int):
            target = min(requested, available)
        else:
            target = requested

        count_name = "full" if requested is None else str(requested)
        default_name = (
            f"{safe_name(args.dataset)}_{safe_name(selected['config'])}_"
            f"{safe_name(selected['split'])}_{count_name}.jsonl"
        )
        output = Path(args.output) if args.output else Path.cwd() / default_name
        output.parent.mkdir(parents=True, exist_ok=True)

        target_text = f"{target:,}" if isinstance(target, int) else "未知总数"
        print(f"\n开始下载：目标 {target_text} 条，单次最多请求 {PAGE_SIZE} 条。")

        written = 0
        with output.open("w", encoding="utf-8") as file:
            while target is None or written < target:
                length = PAGE_SIZE if target is None else min(PAGE_SIZE, target - written)
                print(f"正在请求第 {written + 1} 到 {written + length} 条……", flush=True)
                rows_response = request_json(
                    "/rows",
                    {
                        "dataset": args.dataset,
                        "config": selected["config"],
                        "split": selected["split"],
                        "offset": written,
                        "length": length,
                    },
                    args.timeout,
                )
                rows = rows_response.get("rows", [])
                if not rows:
                    break
                for item in rows:
                    file.write(json.dumps(item.get("row", item), ensure_ascii=False, default=str) + "\n")
                written += len(rows)
                print(f"下载进度：{written:,}/{target_text} 条", flush=True)
                if len(rows) < length:
                    break
    except KeyboardInterrupt:
        print("\n已停止。", file=sys.stderr)
        return 130
    except Exception as error:
        print(f"下载失败：{error}", file=sys.stderr)
        return 1

    print(f"完成：共写入 {written:,} 条")
    print(f"文件：{output.resolve()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: hf_dataset_download.go
// hf_dataset_download is a dependency-free Hugging Face Dataset Viewer downloader.
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
	"unicode"
)

const (
	apiBase  = "https://datasets-server.huggingface.co"
	pageSize = 100 // Dataset Viewer API 每次最多返回 100 条。
)

type splitInfo struct {
	Config  string `json:"config"`
	Split   string `json:"split"`
	NumRows *int64 `json:"num_rows"`
}

type splitsResponse struct {
	Splits []splitInfo `json:"splits"`
}

type rowItem struct {
	Row json.RawMessage `json:"row"`
}

type rowsResponse struct {
	Rows []rowItem `json:"rows"`
}

func main() {
	config := flag.String("config", "", "只显示指定配置（subset）下的 split")
	output := flag.String("o", "", "输出 JSONL 文件路径")
	flag.StringVar(output, "output", "", "输出 JSONL 文件路径")
	timeout := flag.Int("timeout", 60, "单次网络请求超时秒数")
	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "用法：%s [选项] <数据集 ID>\n\n", os.Args[0])
		flag.PrintDefaults()
	}
	flag.Parse()
	if flag.NArg() != 1 {
		flag.Usage()
		os.Exit(2)
	}

	if err := run(flag.Arg(0), *config, *output, time.Duration(*timeout)*time.Second); err != nil {
		fmt.Fprintln(os.Stderr, "下载失败：", err)
		os.Exit(1)
	}
}

func run(dataset, config, output string, timeout time.Duration) error {
	client := &http.Client{Timeout: timeout}
	var response splitsResponse
	if err := getJSON(client, "/splits", url.Values{"dataset": {dataset}}, &response); err != nil {
		return err
	}

	splits := make([]splitInfo, 0, len(response.Splits))
	for _, item := range response.Splits {
		if config == "" || item.Config == config {
			splits = append(splits, item)
		}
	}
	if len(splits) == 0 {
		if config == "" {
			return fmt.Errorf("未找到可用 split；该数据集可能未启用 Dataset Viewer")
		}
		return fmt.Errorf("config %q 下没有可用 split", config)
	}

	labels := make([]string, len(splits))
	for i, item := range splits {
		count := ""
		if item.NumRows != nil {
			count = fmt.Sprintf("，%d 条", *item.NumRows)
		}
		labels[i] = fmt.Sprintf("%s / %s%s", item.Config, item.Split, count)
	}
	selectedIndex, err := choose("可用 config / split：", labels, 0)
	if err != nil {
		return err
	}
	selected := splits[selectedIndex]

	amounts := []string{"100 条", "1k 条", "全部"}
	amountIndex, err := choose("下载数量：", amounts, 0)
	if err != nil {
		return err
	}
	requested := int64(100)
	full := amountIndex == 2
	if amountIndex == 1 {
		requested = 1000
	}

	// target < 0 表示 Viewer 未提供总行数，下载到服务端返回空页为止。
	target := requested
	if full {
		target = -1
		if selected.NumRows != nil {
			target = *selected.NumRows
		}
	} else if selected.NumRows != nil && *selected.NumRows < target {
		target = *selected.NumRows
	}

	if output == "" {
		countName := strconv.FormatInt(requested, 10)
		if full {
			countName = "full"
		}
		output = fmt.Sprintf("%s_%s_%s_%s.jsonl", safeName(dataset), safeName(selected.Config), safeName(selected.Split), countName)
	}
	if err := os.MkdirAll(filepath.Dir(output), 0o755); err != nil && filepath.Dir(output) != "." {
		return err
	}

	targetText := "未知总数"
	if target >= 0 {
		targetText = strconv.FormatInt(target, 10)
	}
	fmt.Printf("\n开始下载：目标 %s 条，单次最多请求 %d 条。\n", targetText, pageSize)

	file, err := os.Create(output)
	if err != nil {
		return err
	}
	defer file.Close()
	writer := bufio.NewWriter(file)
	defer writer.Flush()

	var written int64
	for target < 0 || written < target {
		length := int64(pageSize)
		if target >= 0 && target-written < length {
			length = target - written
		}
		fmt.Printf("正在请求第 %d 到 %d 条……\n", written+1, written+length)

		var page rowsResponse
		err := getJSON(client, "/rows", url.Values{
			"dataset": {dataset},
			"config":  {selected.Config},
			"split":   {selected.Split},
			"offset":  {strconv.FormatInt(written, 10)},
			"length":  {strconv.FormatInt(length, 10)},
		}, &page)
		if err != nil {
			if written != 0 {
				return err
			}
			fmt.Printf("Dataset Viewer 不可用，尝试直接读取原始 JSONL……\n")
			fallbackWritten, rawPath, fallbackErr := downloadRawJSONL(client, dataset, selected, target, writer)
			if fallbackErr != nil {
				return fmt.Errorf("Dataset Viewer 失败：%v；原始 JSONL 回退也失败：%w", err, fallbackErr)
			}
			written = fallbackWritten
			fmt.Printf("已从原始文件 %s 写入 %d 条\n", rawPath, written)
			break
		}
		if len(page.Rows) == 0 {
			break
		}
		for _, item := range page.Rows {
			if _, err := writer.Write(item.Row); err != nil {
				return err
			}
			if err := writer.WriteByte('\n'); err != nil {
				return err
			}
		}
		written += int64(len(page.Rows))
		fmt.Printf("下载进度：%d/%s 条\n", written, targetText)
		if int64(len(page.Rows)) < length {
			break
		}
	}

	if err := writer.Flush(); err != nil {
		return err
	}
	if err := file.Close(); err != nil {
		return err
	}
	fmt.Printf("完成：共写入 %d 条\n文件：%s\n", written, absolutePath(output))
	return nil
}

func choose(prompt string, options []string, defaultIndex int) (int, error) {
	reader := bufio.NewReader(os.Stdin)
	fmt.Printf("\n%s\n", prompt)
	for i, option := range options {
		defaultLabel := ""
		if i == defaultIndex {
			defaultLabel = "（默认）"
		}
		fmt.Printf("  %d. %s%s\n", i+1, option, defaultLabel)
	}

	for {
		fmt.Printf("请输入编号 [默认 %d]: ", defaultIndex+1)
		text, err := reader.ReadString('\n')
		if err == io.EOF && strings.TrimSpace(text) == "" {
			return defaultIndex, nil
		}
		if err != nil && err != io.EOF {
			return 0, err
		}
		text = strings.TrimSpace(text)
		if text == "" {
			return defaultIndex, nil
		}
		index, err := strconv.Atoi(text)
		if err == nil && index >= 1 && index <= len(options) {
			return index - 1, nil
		}
		fmt.Println("输入无效，请输入列表中的编号。")
	}
}

func getJSON(client *http.Client, path string, values url.Values, destination any) error {
	req, err := http.NewRequest(http.MethodGet, apiBase+path+"?"+values.Encode(), nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", "hf-dataset-sample-downloader/1.0")
	if token := hfToken(); token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return fmt.Errorf("HTTP %s: %s", resp.Status, strings.TrimSpace(string(body)))
	}
	return json.NewDecoder(resp.Body).Decode(destination)
}

// downloadRawJSONL is a fallback for datasets whose Dataset Viewer cannot build
// a preview. It reads the source stream and closes the HTTP response as soon as
// the requested number of JSONL records has been written.
func downloadRawJSONL(client *http.Client, dataset string, selected splitInfo, target int64, writer *bufio.Writer) (int64, string, error) {
	paths := []string{
		"data/" + selected.Split + ".jsonl",
		selected.Split + ".jsonl",
		selected.Config + "/" + selected.Split + ".jsonl",
	}

	var lastError error
	for _, path := range paths {
		response, err := openRawFile(client, dataset, path)
		if err != nil {
			lastError = err
			continue
		}
		defer response.Body.Close()

		reader := bufio.NewReader(response.Body)
		var written int64
		for target < 0 || written < target {
			line, readErr := reader.ReadBytes('\n')
			if len(line) > 0 {
				line = bytesTrimLineEnding(line)
				if len(line) > 0 {
					if _, err := writer.Write(line); err != nil {
						return written, path, err
					}
					if err := writer.WriteByte('\n'); err != nil {
						return written, path, err
					}
					written++
					if written == 1 || written%100 == 0 || (target >= 0 && written == target) {
						fmt.Printf("下载进度：%d 条\n", written)
					}
				}
			}
			if readErr == io.EOF {
				break
			}
			if readErr != nil {
				return written, path, readErr
			}
		}
		return written, path, nil
	}
	if lastError == nil {
		lastError = fmt.Errorf("未找到 JSONL 原始文件")
	}
	return 0, "", lastError
}

func openRawFile(client *http.Client, dataset, path string) (*http.Response, error) {
	parts := strings.Split(dataset, "/")
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return nil, fmt.Errorf("无效的数据集 ID：%q", dataset)
	}
	urlPath := "https://huggingface.co/datasets/" + url.PathEscape(parts[0]) + "/" + url.PathEscape(parts[1]) + "/resolve/main/"
	for _, part := range strings.Split(path, "/") {
		urlPath += url.PathEscape(part) + "/"
	}
	urlPath = strings.TrimSuffix(urlPath, "/")

	req, err := http.NewRequest(http.MethodGet, urlPath, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "hf-dataset-sample-downloader/1.0")
	if token := hfToken(); token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}
	response, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
		body, _ := io.ReadAll(io.LimitReader(response.Body, 1024))
		response.Body.Close()
		return nil, fmt.Errorf("%s：HTTP %s: %s", path, response.Status, strings.TrimSpace(string(body)))
	}
	return response, nil
}

func bytesTrimLineEnding(value []byte) []byte {
	return bytes.TrimRight(value, "\r\n")
}

func hfToken() string {
	if token := strings.TrimSpace(os.Getenv("HF_TOKEN")); token != "" {
		return token
	}
	path := os.Getenv("HF_TOKEN_PATH")
	if path == "" {
		cacheRoot := os.Getenv("XDG_CACHE_HOME")
		if cacheRoot == "" {
			home, err := os.UserHomeDir()
			if err != nil {
				return ""
			}
			cacheRoot = filepath.Join(home, ".cache")
		}
		hfHome := os.Getenv("HF_HOME")
		if hfHome == "" {
			hfHome = filepath.Join(cacheRoot, "huggingface")
		}
		path = filepath.Join(hfHome, "token")
	}
	token, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(token))
}

func safeName(text string) string {
	return strings.Map(func(r rune) rune {
		if unicode.IsLetter(r) || unicode.IsDigit(r) || strings.ContainsRune("._-", r) {
			return r
		}
		return '_'
	}, text)
}

func absolutePath(path string) string {
	abs, err := filepath.Abs(path)
	if err != nil {
		return path
	}
	return abs
}
